{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "31e8c3cd",
   "metadata": {},
   "source": [
    "# <span style=\"color:gray\">Pangolin SARS-CoV-2 Pipeline Notebook</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56a29212",
   "metadata": {},
   "source": [
    "We are going to run a standard covid bioinformatics pipeline using the Pangolin workflow. https://cov-lineages.org/resources/pangolin/usage.html"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03541941",
   "metadata": {},
   "source": [
    "### Required software"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f994b990",
   "metadata": {},
   "outputs": [],
   "source": [
    "#change this depending on how many threads are available in your notebook\n",
    "CPU=4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a19b662e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PREFIX=/home/ec2-user/mambaforge\n",
      "Unpacking payload ...\n",
      "Extracting \"libmambapy-0.19.0-py39h8bfa403_0.tar.bz2\"\n",
      "Extracting \"zstd-1.5.0-ha95c52a_0.tar.bz2\"\n",
      "Extracting \"readline-8.1-h46c0cb4_0.tar.bz2\"\n",
      "Extracting \"yaml-cpp-0.6.3-he1b5a44_4.tar.bz2\"\n",
      "Extracting \"libgcc-ng-11.2.0-h1d223b6_11.tar.bz2\"\n",
      "Extracting \"cffi-1.15.0-py39h4bc2ebd_0.tar.bz2\"\n",
      "Extracting \"wheel-0.37.0-pyhd8ed1ab_1.tar.bz2\"\n",
      "Extracting \"colorama-0.4.4-pyh9f0ad1d_0.tar.bz2\"\n",
      "Extracting \"_openmp_mutex-4.5-1_gnu.tar.bz2\"\n",
      "Extracting \"tzdata-2021e-he74cb21_0.tar.bz2\"\n",
      "Extracting \"reproc-cpp-14.2.3-h9c3ff4c_0.tar.bz2\"\n",
      "Extracting \"lz4-c-1.9.3-h9c3ff4c_1.tar.bz2\"\n",
      "Extracting \"libarchive-3.5.2-hccf745f_1.tar.bz2\"\n",
      "Extracting \"libedit-3.1.20191231-he28a2e2_2.tar.bz2\"\n",
      "Extracting \"icu-69.1-h9c3ff4c_0.tar.bz2\"\n",
      "Extracting \"ld_impl_linux-64-2.36.1-hea4e1c9_2.tar.bz2\"\n",
      "Extracting \"libzlib-1.2.11-h36c2ea0_1013.tar.bz2\"\n",
      "Extracting \"tqdm-4.62.3-pyhd8ed1ab_0.tar.bz2\"\n",
      "Extracting \"ncurses-6.2-h58526e2_4.tar.bz2\"\n",
      "Extracting \"openssl-1.1.1l-h7f98852_0.tar.bz2\"\n",
      "Extracting \"cryptography-36.0.0-py39h95dcef6_0.tar.bz2\"\n",
      "Extracting \"zlib-1.2.11-h36c2ea0_1013.tar.bz2\"\n",
      "Extracting \"lzo-2.10-h516909a_1000.tar.bz2\"\n",
      "Extracting \"c-ares-1.18.1-h7f98852_0.tar.bz2\"\n",
      "Extracting \"pyopenssl-21.0.0-pyhd8ed1ab_0.tar.bz2\"\n",
      "Extracting \"conda-package-handling-1.7.3-py39h3811e60_1.tar.bz2\"\n",
      "Extracting \"idna-3.1-pyhd3deb0d_0.tar.bz2\"\n",
      "Extracting \"libmamba-0.19.0-h3985d26_0.tar.bz2\"\n",
      "Extracting \"reproc-14.2.3-h7f98852_0.tar.bz2\"\n",
      "Extracting \"pip-21.3.1-pyhd8ed1ab_0.tar.bz2\"\n",
      "Extracting \"tk-8.6.11-h27826a3_1.tar.bz2\"\n",
      "Extracting \"conda-4.11.0-py39hf3d152e_0.tar.bz2\"\n",
      "Extracting \"requests-2.26.0-pyhd8ed1ab_1.tar.bz2\"\n",
      "Extracting \"_libgcc_mutex-0.1-conda_forge.tar.bz2\"\n",
      "Extracting \"brotlipy-0.7.0-py39h3811e60_1003.tar.bz2\"\n",
      "Extracting \"python-3.9.7-hb7a2778_3_cpython.tar.bz2\"\n",
      "Extracting \"yaml-0.2.5-h516909a_0.tar.bz2\"\n",
      "Extracting \"bzip2-1.0.8-h7f98852_4.tar.bz2\"\n",
      "Extracting \"libffi-3.4.2-h7f98852_5.tar.bz2\"\n",
      "Extracting \"krb5-1.19.2-hcc1bbae_3.tar.bz2\"\n",
      "Extracting \"charset-normalizer-2.0.9-pyhd8ed1ab_0.tar.bz2\"\n",
      "Extracting \"pysocks-1.7.1-py39hf3d152e_4.tar.bz2\"\n",
      "Extracting \"libgomp-11.2.0-h1d223b6_11.tar.bz2\"\n",
      "Extracting \"pybind11-abi-4-hd8ed1ab_3.tar.bz2\"\n",
      "Extracting \"python_abi-3.9-2_cp39.tar.bz2\"\n",
      "Extracting \"libiconv-1.16-h516909a_0.tar.bz2\"\n",
      "Extracting \"libcurl-7.80.0-h2574ce0_0.tar.bz2\"\n",
      "Extracting \"libxml2-2.9.12-h885dcf4_1.tar.bz2\"\n",
      "Extracting \"pycosat-0.6.3-py39h3811e60_1009.tar.bz2\"\n",
      "Extracting \"certifi-2021.10.8-py39hf3d152e_1.tar.bz2\"\n",
      "Extracting \"libssh2-1.10.0-ha56f1ee_2.tar.bz2\"\n",
      "Extracting \"libnghttp2-1.43.0-h812cca2_1.tar.bz2\"\n",
      "Extracting \"mamba-0.19.0-py39hfa8f2c8_0.tar.bz2\"\n",
      "Extracting \"ruamel_yaml-0.15.80-py39h3811e60_1006.tar.bz2\"\n",
      "Extracting \"xz-5.2.5-h516909a_1.tar.bz2\"\n",
      "Extracting \"setuptools-59.4.0-py39hf3d152e_0.tar.bz2\"\n",
      "Extracting \"six-1.16.0-pyh6c4a22f_0.tar.bz2\"\n",
      "Extracting \"urllib3-1.26.7-pyhd8ed1ab_0.tar.bz2\"\n",
      "Extracting \"libstdcxx-ng-11.2.0-he4da1e4_11.tar.bz2\"\n",
      "Extracting \"libsolv-0.7.19-h780b84a_5.tar.bz2\"\n",
      "Extracting \"pycparser-2.21-pyhd8ed1ab_0.tar.bz2\"\n",
      "Extracting \"ca-certificates-2021.10.8-ha878542_0.tar.bz2\"\n",
      "Extracting \"sqlite-3.37.0-h9cd32fc_0.tar.bz2\"\n",
      "Extracting \"libev-4.33-h516909a_1.tar.bz2\"\n",
      "\n",
      "                                           __\n",
      "          __  ______ ___  ____ _____ ___  / /_  ____ _\n",
      "         / / / / __ `__ \\/ __ `/ __ `__ \\/ __ \\/ __ `/\n",
      "        / /_/ / / / / / / /_/ / / / / / / /_/ / /_/ /\n",
      "       / .___/_/ /_/ /_/\\__,_/_/ /_/ /_/_.___/\\__,_/\n",
      "      /_/\n",
      "\n",
      "\n",
      "Transaction\n",
      "\n",
      "  Prefix: /home/ec2-user/mambaforge\n",
      "\n",
      "  Updating specs:\n",
      "\n",
      "   - python==3.9.7=hb7a2778_3_cpython\n",
      "   - _libgcc_mutex==0.1=conda_forge\n",
      "   - ca-certificates==2021.10.8=ha878542_0\n",
      "   - ld_impl_linux-64==2.36.1=hea4e1c9_2\n",
      "   - libstdcxx-ng==11.2.0=he4da1e4_11\n",
      "   - pybind11-abi==4=hd8ed1ab_3\n",
      "   - tzdata==2021e=he74cb21_0\n",
      "   - libgomp==11.2.0=h1d223b6_11\n",
      "   - _openmp_mutex==4.5=1_gnu\n",
      "   - libgcc-ng==11.2.0=h1d223b6_11\n",
      "   - bzip2==1.0.8=h7f98852_4\n",
      "   - c-ares==1.18.1=h7f98852_0\n",
      "   - icu==69.1=h9c3ff4c_0\n",
      "   - libev==4.33=h516909a_1\n",
      "   - libffi==3.4.2=h7f98852_5\n",
      "   - libiconv==1.16=h516909a_0\n",
      "   - libzlib==1.2.11=h36c2ea0_1013\n",
      "   - lz4-c==1.9.3=h9c3ff4c_1\n",
      "   - lzo==2.10=h516909a_1000\n",
      "   - ncurses==6.2=h58526e2_4\n",
      "   - openssl==1.1.1l=h7f98852_0\n",
      "   - reproc==14.2.3=h7f98852_0\n",
      "   - xz==5.2.5=h516909a_1\n",
      "   - yaml==0.2.5=h516909a_0\n",
      "   - yaml-cpp==0.6.3=he1b5a44_4\n",
      "   - libedit==3.1.20191231=he28a2e2_2\n",
      "   - readline==8.1=h46c0cb4_0\n",
      "   - reproc-cpp==14.2.3=h9c3ff4c_0\n",
      "   - zlib==1.2.11=h36c2ea0_1013\n",
      "   - libnghttp2==1.43.0=h812cca2_1\n",
      "   - libsolv==0.7.19=h780b84a_5\n",
      "   - libssh2==1.10.0=ha56f1ee_2\n",
      "   - libxml2==2.9.12=h885dcf4_1\n",
      "   - sqlite==3.37.0=h9cd32fc_0\n",
      "   - tk==8.6.11=h27826a3_1\n",
      "   - zstd==1.5.0=ha95c52a_0\n",
      "   - krb5==1.19.2=hcc1bbae_3\n",
      "   - libarchive==3.5.2=hccf745f_1\n",
      "   - charset-normalizer==2.0.9=pyhd8ed1ab_0\n",
      "   - colorama==0.4.4=pyh9f0ad1d_0\n",
      "   - idna==3.1=pyhd3deb0d_0\n",
      "   - libcurl==7.80.0=h2574ce0_0\n",
      "   - pycparser==2.21=pyhd8ed1ab_0\n",
      "   - python_abi==3.9=2_cp39\n",
      "   - six==1.16.0=pyh6c4a22f_0\n",
      "   - wheel==0.37.0=pyhd8ed1ab_1\n",
      "   - certifi==2021.10.8=py39hf3d152e_1\n",
      "   - cffi==1.15.0=py39h4bc2ebd_0\n",
      "   - libmamba==0.19.0=h3985d26_0\n",
      "   - pycosat==0.6.3=py39h3811e60_1009\n",
      "   - pysocks==1.7.1=py39hf3d152e_4\n",
      "   - ruamel_yaml==0.15.80=py39h3811e60_1006\n",
      "   - setuptools==59.4.0=py39hf3d152e_0\n",
      "   - tqdm==4.62.3=pyhd8ed1ab_0\n",
      "   - brotlipy==0.7.0=py39h3811e60_1003\n",
      "   - conda-package-handling==1.7.3=py39h3811e60_1\n",
      "   - cryptography==36.0.0=py39h95dcef6_0\n",
      "   - libmambapy==0.19.0=py39h8bfa403_0\n",
      "   - pip==21.3.1=pyhd8ed1ab_0\n",
      "   - pyopenssl==21.0.0=pyhd8ed1ab_0\n",
      "   - urllib3==1.26.7=pyhd8ed1ab_0\n",
      "   - requests==2.26.0=pyhd8ed1ab_1\n",
      "   - conda==4.11.0=py39hf3d152e_0\n",
      "   - mamba==0.19.0=py39hfa8f2c8_0\n",
      "\n",
      "\n",
      "  Package                        Version  Build               Channel                                                                        Size\n",
      "───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\n",
      "  Install:\n",
      "───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\n",
      "\n",
      "  + _libgcc_mutex                    0.1  conda_forge         conda-forge/linux-64/_libgcc_mutex-0.1-conda_forge.tar.bz2                   Cached\n",
      "  + _openmp_mutex                    4.5  1_gnu               conda-forge/linux-64/_openmp_mutex-4.5-1_gnu.tar.bz2                         Cached\n",
      "  + brotlipy                       0.7.0  py39h3811e60_1003   conda-forge/linux-64/brotlipy-0.7.0-py39h3811e60_1003.tar.bz2                Cached\n",
      "  + bzip2                          1.0.8  h7f98852_4          conda-forge/linux-64/bzip2-1.0.8-h7f98852_4.tar.bz2                          Cached\n",
      "  + c-ares                        1.18.1  h7f98852_0          conda-forge/linux-64/c-ares-1.18.1-h7f98852_0.tar.bz2                        Cached\n",
      "  + ca-certificates            2021.10.8  ha878542_0          conda-forge/linux-64/ca-certificates-2021.10.8-ha878542_0.tar.bz2            Cached\n",
      "  + certifi                    2021.10.8  py39hf3d152e_1      conda-forge/linux-64/certifi-2021.10.8-py39hf3d152e_1.tar.bz2                Cached\n",
      "  + cffi                          1.15.0  py39h4bc2ebd_0      conda-forge/linux-64/cffi-1.15.0-py39h4bc2ebd_0.tar.bz2                      Cached\n",
      "  + charset-normalizer             2.0.9  pyhd8ed1ab_0        conda-forge/noarch/charset-normalizer-2.0.9-pyhd8ed1ab_0.tar.bz2             Cached\n",
      "  + colorama                       0.4.4  pyh9f0ad1d_0        conda-forge/noarch/colorama-0.4.4-pyh9f0ad1d_0.tar.bz2                       Cached\n",
      "  + conda                         4.11.0  py39hf3d152e_0      conda-forge/linux-64/conda-4.11.0-py39hf3d152e_0.tar.bz2                     Cached\n",
      "  + conda-package-handling         1.7.3  py39h3811e60_1      conda-forge/linux-64/conda-package-handling-1.7.3-py39h3811e60_1.tar.bz2     Cached\n",
      "  + cryptography                  36.0.0  py39h95dcef6_0      conda-forge/linux-64/cryptography-36.0.0-py39h95dcef6_0.tar.bz2              Cached\n",
      "  + icu                             69.1  h9c3ff4c_0          conda-forge/linux-64/icu-69.1-h9c3ff4c_0.tar.bz2                             Cached\n",
      "  + idna                             3.1  pyhd3deb0d_0        conda-forge/noarch/idna-3.1-pyhd3deb0d_0.tar.bz2                             Cached\n",
      "  + krb5                          1.19.2  hcc1bbae_3          conda-forge/linux-64/krb5-1.19.2-hcc1bbae_3.tar.bz2                          Cached\n",
      "  + ld_impl_linux-64              2.36.1  hea4e1c9_2          conda-forge/linux-64/ld_impl_linux-64-2.36.1-hea4e1c9_2.tar.bz2              Cached\n",
      "  + libarchive                     3.5.2  hccf745f_1          conda-forge/linux-64/libarchive-3.5.2-hccf745f_1.tar.bz2                     Cached\n",
      "  + libcurl                       7.80.0  h2574ce0_0          conda-forge/linux-64/libcurl-7.80.0-h2574ce0_0.tar.bz2                       Cached\n",
      "  + libedit                 3.1.20191231  he28a2e2_2          conda-forge/linux-64/libedit-3.1.20191231-he28a2e2_2.tar.bz2                 Cached\n",
      "  + libev                           4.33  h516909a_1          conda-forge/linux-64/libev-4.33-h516909a_1.tar.bz2                           Cached\n",
      "  + libffi                         3.4.2  h7f98852_5          conda-forge/linux-64/libffi-3.4.2-h7f98852_5.tar.bz2                         Cached\n",
      "  + libgcc-ng                     11.2.0  h1d223b6_11         conda-forge/linux-64/libgcc-ng-11.2.0-h1d223b6_11.tar.bz2                    Cached\n",
      "  + libgomp                       11.2.0  h1d223b6_11         conda-forge/linux-64/libgomp-11.2.0-h1d223b6_11.tar.bz2                      Cached\n",
      "  + libiconv                        1.16  h516909a_0          conda-forge/linux-64/libiconv-1.16-h516909a_0.tar.bz2                        Cached\n",
      "  + libmamba                      0.19.0  h3985d26_0          conda-forge/linux-64/libmamba-0.19.0-h3985d26_0.tar.bz2                      Cached\n",
      "  + libmambapy                    0.19.0  py39h8bfa403_0      conda-forge/linux-64/libmambapy-0.19.0-py39h8bfa403_0.tar.bz2                Cached\n",
      "  + libnghttp2                    1.43.0  h812cca2_1          conda-forge/linux-64/libnghttp2-1.43.0-h812cca2_1.tar.bz2                    Cached\n",
      "  + libsolv                       0.7.19  h780b84a_5          conda-forge/linux-64/libsolv-0.7.19-h780b84a_5.tar.bz2                       Cached\n",
      "  + libssh2                       1.10.0  ha56f1ee_2          conda-forge/linux-64/libssh2-1.10.0-ha56f1ee_2.tar.bz2                       Cached\n",
      "  + libstdcxx-ng                  11.2.0  he4da1e4_11         conda-forge/linux-64/libstdcxx-ng-11.2.0-he4da1e4_11.tar.bz2                 Cached\n",
      "  + libxml2                       2.9.12  h885dcf4_1          conda-forge/linux-64/libxml2-2.9.12-h885dcf4_1.tar.bz2                       Cached\n",
      "  + libzlib                       1.2.11  h36c2ea0_1013       conda-forge/linux-64/libzlib-1.2.11-h36c2ea0_1013.tar.bz2                    Cached\n",
      "  + lz4-c                          1.9.3  h9c3ff4c_1          conda-forge/linux-64/lz4-c-1.9.3-h9c3ff4c_1.tar.bz2                          Cached\n",
      "  + lzo                             2.10  h516909a_1000       conda-forge/linux-64/lzo-2.10-h516909a_1000.tar.bz2                          Cached\n",
      "  + mamba                         0.19.0  py39hfa8f2c8_0      conda-forge/linux-64/mamba-0.19.0-py39hfa8f2c8_0.tar.bz2                     Cached\n",
      "  + ncurses                          6.2  h58526e2_4          conda-forge/linux-64/ncurses-6.2-h58526e2_4.tar.bz2                          Cached\n",
      "  + openssl                       1.1.1l  h7f98852_0          conda-forge/linux-64/openssl-1.1.1l-h7f98852_0.tar.bz2                       Cached\n",
      "  + pip                           21.3.1  pyhd8ed1ab_0        conda-forge/noarch/pip-21.3.1-pyhd8ed1ab_0.tar.bz2                           Cached\n",
      "  + pybind11-abi                       4  hd8ed1ab_3          conda-forge/noarch/pybind11-abi-4-hd8ed1ab_3.tar.bz2                         Cached\n",
      "  + pycosat                        0.6.3  py39h3811e60_1009   conda-forge/linux-64/pycosat-0.6.3-py39h3811e60_1009.tar.bz2                 Cached\n",
      "  + pycparser                       2.21  pyhd8ed1ab_0        conda-forge/noarch/pycparser-2.21-pyhd8ed1ab_0.tar.bz2                       Cached\n",
      "  + pyopenssl                     21.0.0  pyhd8ed1ab_0        conda-forge/noarch/pyopenssl-21.0.0-pyhd8ed1ab_0.tar.bz2                     Cached\n",
      "  + pysocks                        1.7.1  py39hf3d152e_4      conda-forge/linux-64/pysocks-1.7.1-py39hf3d152e_4.tar.bz2                    Cached\n",
      "  + python                         3.9.7  hb7a2778_3_cpython  conda-forge/linux-64/python-3.9.7-hb7a2778_3_cpython.tar.bz2                 Cached\n",
      "  + python_abi                       3.9  2_cp39              conda-forge/linux-64/python_abi-3.9-2_cp39.tar.bz2                           Cached\n",
      "  + readline                         8.1  h46c0cb4_0          conda-forge/linux-64/readline-8.1-h46c0cb4_0.tar.bz2                         Cached\n",
      "  + reproc                        14.2.3  h7f98852_0          conda-forge/linux-64/reproc-14.2.3-h7f98852_0.tar.bz2                        Cached\n",
      "  + reproc-cpp                    14.2.3  h9c3ff4c_0          conda-forge/linux-64/reproc-cpp-14.2.3-h9c3ff4c_0.tar.bz2                    Cached\n",
      "  + requests                      2.26.0  pyhd8ed1ab_1        conda-forge/noarch/requests-2.26.0-pyhd8ed1ab_1.tar.bz2                      Cached\n",
      "  + ruamel_yaml                  0.15.80  py39h3811e60_1006   conda-forge/linux-64/ruamel_yaml-0.15.80-py39h3811e60_1006.tar.bz2           Cached\n",
      "  + setuptools                    59.4.0  py39hf3d152e_0      conda-forge/linux-64/setuptools-59.4.0-py39hf3d152e_0.tar.bz2                Cached\n",
      "  + six                           1.16.0  pyh6c4a22f_0        conda-forge/noarch/six-1.16.0-pyh6c4a22f_0.tar.bz2                           Cached\n",
      "  + sqlite                        3.37.0  h9cd32fc_0          conda-forge/linux-64/sqlite-3.37.0-h9cd32fc_0.tar.bz2                        Cached\n",
      "  + tk                            8.6.11  h27826a3_1          conda-forge/linux-64/tk-8.6.11-h27826a3_1.tar.bz2                            Cached\n",
      "  + tqdm                          4.62.3  pyhd8ed1ab_0        conda-forge/noarch/tqdm-4.62.3-pyhd8ed1ab_0.tar.bz2                          Cached\n",
      "  + tzdata                         2021e  he74cb21_0          conda-forge/noarch/tzdata-2021e-he74cb21_0.tar.bz2                           Cached\n",
      "  + urllib3                       1.26.7  pyhd8ed1ab_0        conda-forge/noarch/urllib3-1.26.7-pyhd8ed1ab_0.tar.bz2                       Cached\n",
      "  + wheel                         0.37.0  pyhd8ed1ab_1        conda-forge/noarch/wheel-0.37.0-pyhd8ed1ab_1.tar.bz2                         Cached\n",
      "  + xz                             5.2.5  h516909a_1          conda-forge/linux-64/xz-5.2.5-h516909a_1.tar.bz2                             Cached\n",
      "  + yaml                           0.2.5  h516909a_0          conda-forge/linux-64/yaml-0.2.5-h516909a_0.tar.bz2                           Cached\n",
      "  + yaml-cpp                       0.6.3  he1b5a44_4          conda-forge/linux-64/yaml-cpp-0.6.3-he1b5a44_4.tar.bz2                       Cached\n",
      "  + zlib                          1.2.11  h36c2ea0_1013       conda-forge/linux-64/zlib-1.2.11-h36c2ea0_1013.tar.bz2                       Cached\n",
      "  + zstd                           1.5.0  ha95c52a_0          conda-forge/linux-64/zstd-1.5.0-ha95c52a_0.tar.bz2                           Cached\n",
      "\n",
      "  Summary:\n",
      "\n",
      "  Install: 64 packages\n",
      "\n",
      "  Total download: 0  B\n",
      "\n",
      "───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\n",
      "\n",
      "\n",
      "\n",
      "Transaction starting\n",
      "Linking ca-certificates-2021.10.8-ha878542_0\n",
      "Linking ld_impl_linux-64-2.36.1-hea4e1c9_2\n",
      "Linking libstdcxx-ng-11.2.0-he4da1e4_11\n",
      "Linking pybind11-abi-4-hd8ed1ab_3\n",
      "Linking _libgcc_mutex-0.1-conda_forge\n",
      "Linking tzdata-2021e-he74cb21_0\n",
      "Linking libgomp-11.2.0-h1d223b6_11\n",
      "Linking _openmp_mutex-4.5-1_gnu\n",
      "Linking libgcc-ng-11.2.0-h1d223b6_11\n",
      "Linking ncurses-6.2-h58526e2_4\n",
      "Linking libzlib-1.2.11-h36c2ea0_1013\n",
      "Linking libiconv-1.16-h516909a_0\n",
      "Linking libev-4.33-h516909a_1\n",
      "Linking yaml-cpp-0.6.3-he1b5a44_4\n",
      "Linking yaml-0.2.5-h516909a_0\n",
      "Linking xz-5.2.5-h516909a_1\n",
      "Linking reproc-14.2.3-h7f98852_0\n",
      "Linking openssl-1.1.1l-h7f98852_0\n",
      "Linking lzo-2.10-h516909a_1000\n",
      "Linking lz4-c-1.9.3-h9c3ff4c_1\n",
      "Linking libffi-3.4.2-h7f98852_5\n",
      "Linking icu-69.1-h9c3ff4c_0\n",
      "Linking c-ares-1.18.1-h7f98852_0\n",
      "Linking bzip2-1.0.8-h7f98852_4\n",
      "Linking readline-8.1-h46c0cb4_0\n",
      "Linking libedit-3.1.20191231-he28a2e2_2\n",
      "Linking zlib-1.2.11-h36c2ea0_1013\n",
      "Linking reproc-cpp-14.2.3-h9c3ff4c_0\n",
      "Linking tk-8.6.11-h27826a3_1\n",
      "Linking zstd-1.5.0-ha95c52a_0\n",
      "Linking sqlite-3.37.0-h9cd32fc_0\n",
      "Linking libxml2-2.9.12-h885dcf4_1\n",
      "Linking libssh2-1.10.0-ha56f1ee_2\n",
      "Linking libsolv-0.7.19-h780b84a_5\n",
      "Linking libnghttp2-1.43.0-h812cca2_1\n",
      "Linking krb5-1.19.2-hcc1bbae_3\n",
      "Linking python-3.9.7-hb7a2778_3_cpython\n",
      "Linking libarchive-3.5.2-hccf745f_1\n",
      "Linking libcurl-7.80.0-h2574ce0_0\n",
      "Linking python_abi-3.9-2_cp39\n",
      "Linking wheel-0.37.0-pyhd8ed1ab_1\n",
      "Linking libmamba-0.19.0-h3985d26_0\n",
      "Linking setuptools-59.4.0-py39hf3d152e_0\n",
      "Linking pip-21.3.1-pyhd8ed1ab_0\n",
      "Linking six-1.16.0-pyh6c4a22f_0\n",
      "Linking idna-3.1-pyhd3deb0d_0\n",
      "Linking libmambapy-0.19.0-py39h8bfa403_0\n",
      "Linking ruamel_yaml-0.15.80-py39h3811e60_1006\n",
      "Linking pysocks-1.7.1-py39hf3d152e_4\n",
      "Linking pycosat-0.6.3-py39h3811e60_1009\n",
      "Linking certifi-2021.10.8-py39hf3d152e_1\n",
      "Linking pycparser-2.21-pyhd8ed1ab_0\n",
      "Linking colorama-0.4.4-pyh9f0ad1d_0\n",
      "Linking charset-normalizer-2.0.9-pyhd8ed1ab_0\n",
      "Linking cffi-1.15.0-py39h4bc2ebd_0\n",
      "Linking tqdm-4.62.3-pyhd8ed1ab_0\n",
      "Linking cryptography-36.0.0-py39h95dcef6_0\n",
      "Linking brotlipy-0.7.0-py39h3811e60_1003\n",
      "Linking conda-package-handling-1.7.3-py39h3811e60_1\n",
      "Linking pyopenssl-21.0.0-pyhd8ed1ab_0\n",
      "Linking urllib3-1.26.7-pyhd8ed1ab_0\n",
      "Linking requests-2.26.0-pyhd8ed1ab_1\n",
      "Linking conda-4.11.0-py39hf3d152e_0\n",
      "Linking mamba-0.19.0-py39hfa8f2c8_0\n",
      "Transaction finished\n",
      "installation finished.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   160  100   160    0     0    969      0 --:--:-- --:--:-- --:--:--   969\n",
      "100   665  100   665    0     0   2224      0 --:--:-- --:--:-- --:--:--  2224\n",
      "100  102M  100  102M    0     0  14.3M      0  0:00:07  0:00:07 --:--:-- 14.4M\n",
      "bash: line 7: !cp: command not found\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "#install Mamba\n",
    "curl -L -O https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-$(uname)-$(uname -m).sh\n",
    "bash Mambaforge-$(uname)-$(uname -m).sh -b -p $HOME/mambaforge\n",
    "rm Mamba*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a40f7ebc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#move mamba executable to your path\n",
    "!cp ~/mambaforge/bin/mamba /home/ec2-user/anaconda3/condabin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f421805e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: biopython in /home/ec2-user/anaconda3/envs/amazonei_mxnet_p36/lib/python3.6/site-packages (1.79)\n",
      "Requirement already satisfied: numpy in /home/ec2-user/anaconda3/envs/amazonei_mxnet_p36/lib/python3.6/site-packages (from biopython) (1.19.5)\n",
      "\u001b[33mWARNING: You are using pip version 21.2.4; however, version 21.3.1 is available.\n",
      "You should consider upgrading via the '/home/ec2-user/anaconda3/envs/amazonei_mxnet_p36/bin/python -m pip install --upgrade pip' command.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "#install biopython to import packages below\n",
    "!pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "fd936fd6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: - \n",
      "The environment is inconsistent, please check the package plan carefully\n",
      "The following packages are causing the inconsistency:\n",
      "\n",
      "  - conda-forge/noarch::seaborn-base==0.11.1=pyhd8ed1ab_1\n",
      "  - conda-forge/noarch::nbclassic==0.2.6=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::typing-extensions==3.7.4.3=0\n",
      "  - conda-forge/linux-64::pluggy==0.13.1=py36h5fab9bb_4\n",
      "  - conda-forge/linux-64::blaze==0.11.3=py36_0\n",
      "  - conda-forge/linux-64::matplotlib==3.3.4=py36h5fab9bb_0\n",
      "  - conda-forge/noarch::python-language-server==0.36.2=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::jupyterlab_server==2.3.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::pyls-black==0.4.6=pyh9f0ad1d_0\n",
      "  - conda-forge/linux-64::scikit-image==0.16.2=py36hb3f55d8_0\n",
      "  - conda-forge/noarch::path.py==12.5.0=0\n",
      "  - conda-forge/noarch::qdarkstyle==2.8.1=pyhd8ed1ab_2\n",
      "  - conda-forge/noarch::ipywidgets==7.6.3=pyhd3deb0d_0\n",
      "  - conda-forge/noarch::black==20.8b1=py_1\n",
      "  - conda-forge/linux-64::anyio==2.1.0=py36h5fab9bb_0\n",
      "  - conda-forge/linux-64::jupyter_server==1.4.1=py36h5fab9bb_0\n",
      "  - conda-forge/noarch::nbclient==0.5.2=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::widgetsnbextension==3.5.1=py36h5fab9bb_4\n",
      "  - conda-forge/linux-64::bokeh==2.2.3=py36h5fab9bb_0\n",
      "  - conda-forge/linux-64::keyring==22.0.1=py36h5fab9bb_0\n",
      "  - conda-forge/linux-64::nbconvert==6.0.7=py36h5fab9bb_3\n",
      "  - conda-forge/noarch::numpydoc==1.1.0=py_1\n",
      "  - conda-forge/linux-64::spyder==4.2.0=py36h5fab9bb_0\n",
      "  - conda-forge/noarch::flake8==3.8.4=py_0\n",
      "  - conda-forge/noarch::pyls-spyder==0.3.2=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::nbformat==5.1.2=pyhd8ed1ab_1\n",
      "  - conda-forge/noarch::importlib_metadata==3.6.0=hd8ed1ab_0\n",
      "  - conda-forge/noarch::aioitertools==0.7.1=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::jupyterlab_launcher==0.13.1=py_2\n",
      "  - conda-forge/noarch::odo==0.5.1=py_1\n",
      "  - conda-forge/noarch::imageio==2.9.0=py_0\n",
      "  - conda-forge/noarch::helpdev==0.7.1=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::path==15.1.2=py36h5fab9bb_0\n",
      "  - conda-forge/noarch::jsonschema==3.2.0=py_2\n",
      "  - conda-forge/linux-64::yarl==1.6.3=py36h8f6f2f9_1\n",
      "  - conda-forge/noarch::sphinx==3.5.1=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::seaborn==0.11.1=hd8ed1ab_1\n",
      "  - conda-forge/linux-64::jupyter==1.0.0=py36h5fab9bb_6\n",
      "  - conda-forge/linux-64::nb_conda==2.2.1=py36h5fab9bb_4\n",
      "  - conda-forge/noarch::dask==2021.2.0=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::matplotlib-base==3.3.4=py36hd391965_0\n",
      "  - conda-forge/noarch::anaconda-client==1.7.2=py_0\n",
      "  - conda-forge/noarch::anaconda-project==0.9.1=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::importlib-metadata==3.6.0=py36h5fab9bb_0\n",
      "  - conda-forge/linux-64::pytest==6.2.2=py36h5fab9bb_0\n",
      "failed with initial frozen solve. Retrying with flexible solve.\n",
      "Collecting package metadata (repodata.json): done\n",
      "Solving environment: - \n",
      "The environment is inconsistent, please check the package plan carefully\n",
      "The following packages are causing the inconsistency:\n",
      "\n",
      "  - conda-forge/noarch::seaborn-base==0.11.1=pyhd8ed1ab_1\n",
      "  - conda-forge/noarch::nbclassic==0.2.6=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::typing-extensions==3.7.4.3=0\n",
      "  - conda-forge/linux-64::pluggy==0.13.1=py36h5fab9bb_4\n",
      "  - conda-forge/linux-64::blaze==0.11.3=py36_0\n",
      "  - conda-forge/linux-64::matplotlib==3.3.4=py36h5fab9bb_0\n",
      "  - conda-forge/noarch::python-language-server==0.36.2=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::jupyterlab_server==2.3.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::pyls-black==0.4.6=pyh9f0ad1d_0\n",
      "  - conda-forge/linux-64::scikit-image==0.16.2=py36hb3f55d8_0\n",
      "  - conda-forge/noarch::path.py==12.5.0=0\n",
      "  - conda-forge/noarch::qdarkstyle==2.8.1=pyhd8ed1ab_2\n",
      "  - conda-forge/noarch::ipywidgets==7.6.3=pyhd3deb0d_0\n",
      "  - conda-forge/noarch::black==20.8b1=py_1\n",
      "  - conda-forge/linux-64::anyio==2.1.0=py36h5fab9bb_0\n",
      "  - conda-forge/linux-64::jupyter_server==1.4.1=py36h5fab9bb_0\n",
      "  - conda-forge/noarch::nbclient==0.5.2=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::widgetsnbextension==3.5.1=py36h5fab9bb_4\n",
      "  - conda-forge/linux-64::bokeh==2.2.3=py36h5fab9bb_0\n",
      "  - conda-forge/linux-64::keyring==22.0.1=py36h5fab9bb_0\n",
      "  - conda-forge/linux-64::nbconvert==6.0.7=py36h5fab9bb_3\n",
      "  - conda-forge/noarch::numpydoc==1.1.0=py_1\n",
      "  - conda-forge/linux-64::spyder==4.2.0=py36h5fab9bb_0\n",
      "  - conda-forge/noarch::flake8==3.8.4=py_0\n",
      "  - conda-forge/noarch::pyls-spyder==0.3.2=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::nbformat==5.1.2=pyhd8ed1ab_1\n",
      "  - conda-forge/noarch::importlib_metadata==3.6.0=hd8ed1ab_0\n",
      "  - conda-forge/noarch::aioitertools==0.7.1=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::jupyterlab_launcher==0.13.1=py_2\n",
      "  - conda-forge/noarch::odo==0.5.1=py_1\n",
      "  - conda-forge/noarch::imageio==2.9.0=py_0\n",
      "  - conda-forge/noarch::helpdev==0.7.1=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::path==15.1.2=py36h5fab9bb_0\n",
      "  - conda-forge/noarch::jsonschema==3.2.0=py_2\n",
      "  - conda-forge/linux-64::yarl==1.6.3=py36h8f6f2f9_1\n",
      "  - conda-forge/noarch::sphinx==3.5.1=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::seaborn==0.11.1=hd8ed1ab_1\n",
      "  - conda-forge/linux-64::jupyter==1.0.0=py36h5fab9bb_6\n",
      "  - conda-forge/linux-64::nb_conda==2.2.1=py36h5fab9bb_4\n",
      "  - conda-forge/noarch::dask==2021.2.0=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::matplotlib-base==3.3.4=py36hd391965_0\n",
      "  - conda-forge/noarch::anaconda-client==1.7.2=py_0\n",
      "  - conda-forge/noarch::anaconda-project==0.9.1=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::importlib-metadata==3.6.0=py36h5fab9bb_0\n",
      "  - conda-forge/linux-64::pytest==6.2.2=py36h5fab9bb_0\n",
      "done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 4.8.4\n",
      "  latest version: 4.11.0\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c defaults conda\n",
      "\n",
      "\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/ec2-user/anaconda3/envs/amazonei_mxnet_p36\n",
      "\n",
      "  added / updated specs:\n",
      "    - ipyrad\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    astroid-2.7.3              |   py36h5fab9bb_0         330 KB  conda-forge\n",
      "    bedtools-2.30.0            |       h7d7f7ad_1        17.9 MB  bioconda\n",
      "    cutadapt-3.4               |   py36hc5360cc_1         197 KB  bioconda\n",
      "    dataclasses-0.8            |     pyh787bdff_2          22 KB  conda-forge\n",
      "    dnaio-0.5.1                |   py36hc5360cc_0         137 KB  bioconda\n",
      "    flask-cors-3.0.10          |     pyhd8ed1ab_0          15 KB  conda-forge\n",
      "    fsspec-2021.11.1           |     pyhd8ed1ab_0          91 KB  conda-forge\n",
      "    htslib-1.11                |       hd3b49d5_1         1.7 MB  bioconda\n",
      "    jupyter_console-5.2.0      |           py36_1          34 KB  conda-forge\n",
      "    libdeflate-1.6             |       h516909a_0          60 KB  conda-forge\n",
      "    mpi4py-3.0.3               |   py36he1a1962_7         696 KB  conda-forge\n",
      "    notebook-6.3.0             |   py36h5fab9bb_0         6.3 MB  conda-forge\n",
      "    perl-5.32.1                | 0_h7f98852_perl5        14.5 MB  conda-forge\n",
      "    pillow-8.2.0               |   py36ha6010c0_1         688 KB  conda-forge\n",
      "    platformdirs-2.3.0         |     pyhd8ed1ab_0          14 KB  conda-forge\n",
      "    pylint-2.10.2              |     pyhd8ed1ab_0         255 KB  conda-forge\n",
      "    pysam-0.16.0.1             |   py36h4c34d4e_1         2.5 MB  bioconda\n",
      "    python-isal-0.11.0         |   py36h8f6f2f9_0         136 KB  conda-forge\n",
      "    reportlab-3.5.68           |   py36h3e18861_0         2.4 MB  conda-forge\n",
      "    samtools-1.11              |       h6270b1f_0         383 KB  bioconda\n",
      "    typing_extensions-3.7.4.3  |             py_0          25 KB  conda-forge\n",
      "    vsearch-2.17.1             |       h95f258a_0         1.4 MB  bioconda\n",
      "    xopen-1.2.0                |   py36h5fab9bb_0          22 KB  conda-forge\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:        49.7 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  arrow              conda-forge/noarch::arrow-1.2.1-pyhd8ed1ab_0\n",
      "  astroid            conda-forge/linux-64::astroid-2.7.3-py36h5fab9bb_0\n",
      "  bedtools           bioconda/linux-64::bedtools-2.30.0-h7d7f7ad_1\n",
      "  bwa                bioconda/linux-64::bwa-0.7.17-h5bf99c6_8\n",
      "  charset-normalizer conda-forge/noarch::charset-normalizer-2.0.9-pyhd8ed1ab_0\n",
      "  colorama           conda-forge/noarch::colorama-0.4.4-pyh9f0ad1d_0\n",
      "  custom-inherit     conda-forge/noarch::custom-inherit-2.4.0-pyhd8ed1ab_0\n",
      "  cutadapt           bioconda/linux-64::cutadapt-3.4-py36hc5360cc_1\n",
      "  dataclasses        conda-forge/noarch::dataclasses-0.8-pyh787bdff_2\n",
      "  dnaio              bioconda/linux-64::dnaio-0.5.1-py36hc5360cc_0\n",
      "  docutils           conda-forge/linux-64::docutils-0.16-py36h5fab9bb_3\n",
      "  flask-cors         conda-forge/noarch::flask-cors-3.0.10-pyhd8ed1ab_0\n",
      "  fsspec             conda-forge/noarch::fsspec-2021.11.1-pyhd8ed1ab_0\n",
      "  htslib             bioconda/linux-64::htslib-1.11-hd3b49d5_1\n",
      "  ipyrad             bioconda/noarch::ipyrad-0.9.81-pyh5e36f6f_0\n",
      "  isa-l              conda-forge/linux-64::isa-l-2.30.0-ha770c72_4\n",
      "  jupyter_console    conda-forge/linux-64::jupyter_console-5.2.0-py36_1\n",
      "  libdeflate         conda-forge/linux-64::libdeflate-1.6-h516909a_0\n",
      "  mpi                conda-forge/linux-64::mpi-1.0-openmpi\n",
      "  mpi4py             conda-forge/linux-64::mpi4py-3.0.3-py36he1a1962_7\n",
      "  muscle             bioconda/linux-64::muscle-3.8.1551-h7d875b9_6\n",
      "  notebook           conda-forge/linux-64::notebook-6.3.0-py36h5fab9bb_0\n",
      "  openjpeg           conda-forge/linux-64::openjpeg-2.4.0-hb52868f_1\n",
      "  openmpi            conda-forge/linux-64::openmpi-4.1.1-hbfc84c5_0\n",
      "  pbzip2             conda-forge/linux-64::pbzip2-1.1.13-0\n",
      "  perl               conda-forge/linux-64::perl-5.32.1-0_h7f98852_perl5\n",
      "  pigz               conda-forge/linux-64::pigz-2.6-h27826a3_0\n",
      "  pillow             conda-forge/linux-64::pillow-8.2.0-py36ha6010c0_1\n",
      "  platformdirs       conda-forge/noarch::platformdirs-2.3.0-pyhd8ed1ab_0\n",
      "  pylint             conda-forge/noarch::pylint-2.10.2-pyhd8ed1ab_0\n",
      "  pypng              conda-forge/noarch::pypng-0.0.20-py_0\n",
      "  pysam              bioconda/linux-64::pysam-0.16.0.1-py36h4c34d4e_1\n",
      "  python-isal        conda-forge/linux-64::python-isal-0.11.0-py36h8f6f2f9_0\n",
      "  reportlab          conda-forge/linux-64::reportlab-3.5.68-py36h3e18861_0\n",
      "  requests           conda-forge/noarch::requests-2.26.0-pyhd8ed1ab_1\n",
      "  samtools           bioconda/linux-64::samtools-1.11-h6270b1f_0\n",
      "  toyplot            conda-forge/noarch::toyplot-0.19.0-pyh9f0ad1d_0\n",
      "  typing_extensions  conda-forge/noarch::typing_extensions-3.7.4.3-py_0\n",
      "  urllib3            conda-forge/noarch::urllib3-1.26.7-pyhd8ed1ab_0\n",
      "  vsearch            bioconda/linux-64::vsearch-2.17.1-h95f258a_0\n",
      "  xopen              conda-forge/linux-64::xopen-1.2.0-py36h5fab9bb_0\n",
      "\n",
      "The following packages will be DOWNGRADED:\n",
      "\n",
      "  libgcc-ng                              11.2.0-h1d223b6_11 --> 9.3.0-h2828fa1_18\n",
      "  libgomp                                11.2.0-h1d223b6_11 --> 9.3.0-h2828fa1_18\n",
      "  openssl                                 1.1.1l-h7f98852_0 --> 1.1.1k-h7f98852_0\n",
      "\n",
      "\n",
      "\n",
      "Downloading and Extracting Packages\n",
      "reportlab-3.5.68     | 2.4 MB    | ##################################### | 100% \n",
      "dnaio-0.5.1          | 137 KB    | ##################################### | 100% \n",
      "htslib-1.11          | 1.7 MB    | ##################################### | 100% \n",
      "cutadapt-3.4         | 197 KB    | ##################################### | 100% \n",
      "libdeflate-1.6       | 60 KB     | ##################################### | 100% \n",
      "flask-cors-3.0.10    | 15 KB     | ##################################### | 100% \n",
      "typing_extensions-3. | 25 KB     | ##################################### | 100% \n",
      "samtools-1.11        | 383 KB    | ##################################### | 100% \n",
      "fsspec-2021.11.1     | 91 KB     | ##################################### | 100% \n",
      "bedtools-2.30.0      | 17.9 MB   | ##################################### | 100% \n",
      "perl-5.32.1          | 14.5 MB   | ##################################### | 100% \n",
      "python-isal-0.11.0   | 136 KB    | ##################################### | 100% \n",
      "dataclasses-0.8      | 22 KB     | ##################################### | 100% \n",
      "pillow-8.2.0         | 688 KB    | ##################################### | 100% \n",
      "astroid-2.7.3        | 330 KB    | ##################################### | 100% \n",
      "pylint-2.10.2        | 255 KB    | ##################################### | 100% \n",
      "pysam-0.16.0.1       | 2.5 MB    | ##################################### | 100% \n",
      "vsearch-2.17.1       | 1.4 MB    | ##################################### | 100% \n",
      "jupyter_console-5.2. | 34 KB     | ##################################### | 100% \n",
      "xopen-1.2.0          | 22 KB     | ##################################### | 100% \n",
      "mpi4py-3.0.3         | 696 KB    | ##################################### | 100% \n",
      "platformdirs-2.3.0   | 14 KB     | ##################################### | 100% \n",
      "notebook-6.3.0       | 6.3 MB    | ##################################### | 100% \n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: -  \n",
      "For Linux 64, Open MPI is built with CUDA awareness but this support is disabled by default.\n",
      "To enable it, please set the environment variable OMPI_MCA_opal_cuda_support=true before\n",
      "launching your MPI processes. Equivalently, you can set the MCA parameter in the command line:\n",
      "mpiexec --mca opal_cuda_support 1 ...\n",
      " \n",
      "In addition, the UCX support is also built but disabled by default.\n",
      "To enable it, first install UCX (conda install -c conda-forge ucx). Then, set the environment\n",
      "variables OMPI_MCA_pml=\"ucx\" OMPI_MCA_osc=\"ucx\" before launching your MPI processes.\n",
      "Equivalently, you can set the MCA parameters in the command line:\n",
      "mpiexec --mca pml ucx --mca osc ucx ...\n",
      "Note that you might also need to set UCX_MEMTYPE_CACHE=n for CUDA awareness via UCX.\n",
      "Please consult UCX's documentation for detail.\n",
      " \n",
      "\n",
      "done\n"
     ]
    }
   ],
   "source": [
    "!conda install ipyrad -y -c conda-forge -c bioconda"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d0f27ee",
   "metadata": {},
   "source": [
    "Now we want to create a conda/mamba env that has all of our necessary dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "4ba6fae7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "name: pangolin\n",
      "channels:\n",
      "  - bioconda\n",
      "  - conda-forge\n",
      "  - defaults\n",
      "  - eaton-lab\n",
      "  \n",
      "dependencies:\n",
      "  - sra-tools\n",
      "  - ipyrad\n",
      "  - toytree\n",
      "  - pangolin\n",
      "  - iqtree\n"
     ]
    }
   ],
   "source": [
    "#you can look at the yaml file that specifies which programs we want to install\n",
    "#you can also specify specific versions, here we just use the latest conda versionå\n",
    "#for example, - sra-tools=2.11.0\n",
    "!cat pangolin.yaml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "49a20dc5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: mamba [-h] {create,export,list,remove,update,config} ...\n",
      "mamba: error: unrecognized arguments: -y\n"
     ]
    }
   ],
   "source": [
    "#create the environment. Here we use mamba because it is faster than conda\n",
    "!mamba env create -f pangolin.yaml -y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "fd23abbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#give it the whole path to the env because otherwise it can't find the env\n",
    "#if you want to play with it add a cell and type 'conda activate pangolin' \n",
    "#or 'source activate pangolin'\n",
    "!source activate /home/ec2-user/mambaforge/envs/pangolin\n",
    "#!mamba info --envs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "96dd7966",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "                  __    __    __    __\n",
      "                 /  \\  /  \\  /  \\  /  \\\n",
      "                /    \\/    \\/    \\/    \\\n",
      "███████████████/  /██/  /██/  /██/  /████████████████████████\n",
      "              /  / \\   / \\   / \\   / \\  \\____\n",
      "             /  /   \\_/   \\_/   \\_/   \\    o \\__,\n",
      "            / _/                       \\_____/  `\n",
      "            |/\n",
      "        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗\n",
      "        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗\n",
      "        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║\n",
      "        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║\n",
      "        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║\n",
      "        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝\n",
      "\n",
      "        mamba (0.19.0) supported by @QuantStack\n",
      "\n",
      "        GitHub:  https://github.com/mamba-org/mamba\n",
      "        Twitter: https://twitter.com/QuantStack\n",
      "\n",
      "█████████████████████████████████████████████████████████████\n",
      "\n",
      "\n",
      "Looking for: ['iqtree']\n",
      "\n",
      "bioconda/linux-64        Using cache\n",
      "bioconda/noarch          Using cache\n",
      "conda-forge/linux-64     Using cache\n",
      "conda-forge/noarch       Using cache\n",
      "pkgs/main/linux-64       Using cache\n",
      "pkgs/main/noarch         Using cache\n",
      "pkgs/r/linux-64          Using cache\n",
      "pkgs/r/noarch            Using cache\n",
      "\n",
      "Pinned packages:\n",
      "  - python 3.6.*\n",
      "\n",
      "\n",
      "Transaction\n",
      "\n",
      "  Prefix: /home/ec2-user/anaconda3/envs/amazonei_mxnet_p36\n",
      "\n",
      "  All requested packages already installed\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!mamba install -c bioconda iqtree -y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "5a99cf0d",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'ipyrad'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-40-a282b1625a17>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mBio\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mSeqIO\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mBio\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mEntrez\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mipyrad\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0manalysis\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mipa\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mtoytree\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'ipyrad'"
     ]
    }
   ],
   "source": [
    "#import libraries\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "import ipyrad.analysis as ipa\n",
    "import toytree"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc694629",
   "metadata": {},
   "source": [
    "### Set up your directory structure and remove files from previous runs if they exist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "0f0e81f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/jupyter/cloud-lab-training/GCP/notebooks/pangolin\n"
     ]
    }
   ],
   "source": [
    "cd /home/jupyter/cloud-lab-training/GCP/notebooks/pangolin/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "8f831fca",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists('pangolin_analysis'):\n",
    "    os.mkdir('pangolin_analysis')\n",
    "os.chdir('pangolin_analysis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "6423ca5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rm: cannot remove 'sarscov2_*': No such file or directory\n",
      "rm: cannot remove 'lineage_report.csv': No such file or directory\n"
     ]
    }
   ],
   "source": [
    "if os.path.exists('sarscov2_sequences.fasta'):\n",
    "    os.remove('sarscov2_sequences.fasta')\n",
    "!rm sarscov2_*\n",
    "!rm lineage_report.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d7015e6",
   "metadata": {},
   "source": [
    "### Fetch viral sequences using a list of accession IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "16824bcf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the number of sequences we will analyze =  18\n"
     ]
    }
   ],
   "source": [
    "#give a list of accession number for covid sequences\n",
    "acc_nums=['NC_045512','LR757995','LR757996','OL698718','OL677199','OL672836','MZ914912','MZ916499','MZ908464','MW580573','MW580574','MW580576','MW991906','MW931310','MW932027','MW424864','MW453109','MW453110']\n",
    "print('the number of sequences we will analyze = ',len(acc_nums))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e382d33",
   "metadata": {},
   "source": [
    "Let this block run without going to the next until it finishes, otherwise you may get an error about too many requests. If that happens, reset your kernel and just rerun everything (except installing software)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a28a7122",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved NC_045512\n",
      "Saved LR757995\n",
      "Saved LR757996\n",
      "Saved OL698718\n",
      "Saved OL677199\n",
      "Saved OL672836\n",
      "Saved MZ914912\n",
      "Saved MZ916499\n",
      "Saved MZ908464\n",
      "Saved MW580573\n",
      "Saved MW580574\n",
      "Saved MW580576\n",
      "Saved MW991906\n",
      "Saved MW931310\n",
      "Saved MW932027\n",
      "Saved MW424864\n",
      "Saved MW453109\n",
      "Saved MW453110\n"
     ]
    }
   ],
   "source": [
    "#use the bio.entrez toolkit within biopython to download the accession numbers\n",
    "#save those sequences to a single fasta file\n",
    "Entrez.email = \"email@example.com\"  # Always tell NCBI who you are\n",
    "filename = \"sarscov2_seqs.fasta\"\n",
    "if not os.path.isfile(filename):\n",
    "    # Downloading...\n",
    "    for acc in acc_nums:\n",
    "        net_handle = Entrez.efetch(\n",
    "            db=\"nucleotide\", id=acc, rettype=\"fasta\", retmode=\"text\"\n",
    "        )\n",
    "        out_handle = open(filename, \"a\")\n",
    "        out_handle.write(net_handle.read())\n",
    "        out_handle.close()\n",
    "        net_handle.close()\n",
    "        print(\"Saved\",acc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "56acb7cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the number of seqs in our fasta file: \n",
      "18\n"
     ]
    }
   ],
   "source": [
    "#make sure our fasta file has the same number of seqs as the acc_nums list\n",
    "print('the number of seqs in our fasta file: ')\n",
    "!grep '>' sarscov2_seqs.fasta | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8606c352",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">NC_045512.2 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome\n",
      "ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGTAGATCTGTTCTCTAAA\n",
      "CGAACTTTAAAATCTGTGTGGCTGTCACTCGGCTGCATGCTTAGTGCACTCACGCAGTATAATTAATAAC\n",
      "TAATTACTGTCGTTGACAGGACACGAGTAACTCGTCTATCTTCTGCAGGCTGCTTACGGTTTCGTCCGTG\n",
      "TTGCAGCCGATCATCAGCACATCTAGGTTTCGTCCGGGTGTGACCGAAAGGTAAGATGGAGAGCCTTGTC\n",
      "CCTGGTTTCAACGAGAAAACACACGTCCAACTCAGTTTGCCTGTTTTACAGGTTCGCGACGTGCTCGTAC\n",
      "GTGGCTTTGGAGACTCCGTGGAGGAGGTCTTATCAGAGGCACGTCAACATCTTAAAGATGGCACTTGTGG\n",
      "CTTAGTAGAAGTTGAAAAAGGCGTTTTGCCTCAACTTGAACAGCCCTATGTGTTCATCAAACGTTCGGAT\n",
      "GCTCGAACTGCACCTCATGGTCATGTTATGGTTGAGCTGGTAGCAGAACTCGAAGGCATTCAGTACGGTC\n",
      "GTAGTGGTGAGACACTTGGTGTCCTTGTCCCTCATGTGGGCGAAATACCAGTGGCTTACCGCAAGGTTCT\n"
     ]
    }
   ],
   "source": [
    "#let's peek at our new fasta file\n",
    "!head sarscov2_seqs.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2db37b4e",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Run pangolin to identify lineages and output alignment\n",
    "Here we call pangolin, give it our input sequences and the number of threads. We also tell it to output the alignment. The full list of pangolin parameters can be found in the [docs](https://cov-lineages.org/resources/pangolin/usage.html)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "f1a17a74",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mAll dependencies satisfied.\u001b[0m\n",
      "\u001b[32mThe query file is:\u001b[0m/home/jupyter/cloud-lab-training/GCP/notebooks/pangolin/pangolin_analysis/sarscov2_seqs.fasta\n",
      "\u001b[32m** Running sequence QC **\u001b[0m\n",
      "\u001b[32mNumber of sequences detected: \u001b[0m18\n",
      "\u001b[32mTotal passing QC: \u001b[0m18\n",
      "\u001b[32m\n",
      "Data files found:\u001b[0m\n",
      "Trained model:\t/opt/conda/lib/python3.7/site-packages/pangoLEARN/data/decisionTree_v1.joblib\n",
      "Header file:\t/opt/conda/lib/python3.7/site-packages/pangoLEARN/data/decisionTreeHeaders_v1.joblib\n",
      "Designated hash:\t/opt/conda/lib/python3.7/site-packages/pangoLEARN/data/lineages.hash.csv\n",
      "\u001b[33mJob stats:\n",
      "job                     count    min threads    max threads\n",
      "--------------------  -------  -------------  -------------\n",
      "add_failed_seqs             1              1              1\n",
      "align_to_reference          1              1              1\n",
      "all                         1              1              1\n",
      "generate_report             1              1              1\n",
      "get_constellations          1              1              1\n",
      "hash_sequence_assign        1              1              1\n",
      "pangolearn                  1              1              1\n",
      "scorpio                     1              4              4\n",
      "total                       8              1              4\n",
      "\u001b[0m\n",
      "loading model 12/04/2021, 00:00:50\n",
      "/opt/conda/lib/python3.7/site-packages/sklearn/base.py:334: UserWarning: Trying to unpickle estimator DecisionTreeClassifier from version 0.24.2 when using version 0.23.1. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  UserWarning)\n",
      "processing block of 6 sequences 12/04/2021, 00:00:51\n",
      "complete 12/04/2021, 00:00:51\n",
      "\u001b[32mOutput file written to: \u001b[0m/home/jupyter/cloud-lab-training/GCP/notebooks/pangolin/pangolin_analysis/lineage_report.csv\n",
      "\u001b[32mOutput alignment written to: \u001b[0m/home/jupyter/cloud-lab-training/GCP/notebooks/pangolin/pangolin_analysis/sequences.aln.fasta\n"
     ]
    }
   ],
   "source": [
    "!pangolin sarscov2_seqs.fasta --alignment --threads $CPU"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0e56a4b",
   "metadata": {},
   "source": [
    "You can view the output file from pangolin called lineage_report.csv (within pangolin_analysis folder) by double clicking on the file, or by right clicking and downloading. What lineages are present in the dataset? Is Omicron in there?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37e6efbe",
   "metadata": {},
   "source": [
    "### Run iqtree to estimate maximum likelihood tree for our sequences\n",
    "iqtree can find the best nucleotide model for the data, but here we are going to assign a model to save time (HKY) and just estimate the phylogeny without any bootstrap support values. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f2782855",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ-TREE multicore version 2.1.4-beta COVID-edition for Linux 64-bit built Jun 24 2021\n",
      "Developed by Bui Quang Minh, James Barbetti, Nguyen Lam Tung,\n",
      "Olga Chernomor, Heiko Schmidt, Dominik Schrempf, Michael Woodhams.\n",
      "\n",
      "Host:    cloud-lab-notebook (AVX2, FMA3, 14 GB RAM)\n",
      "Command: iqtree -s sequences.aln.fasta -nt 4 -m HKY --prefix sarscov2_tree --redo-tree\n",
      "Seed:    719057 (Using SPRNG - Scalable Parallel Random Number Generator)\n",
      "Time:    Fri Dec  3 23:53:05 2021\n",
      "Kernel:  AVX+FMA - 4 threads (4 CPU cores detected)\n",
      "\n",
      "Reading alignment file sequences.aln.fasta ... Fasta format detected\n",
      "Alignment most likely contains DNA/RNA sequences\n",
      "WARNING: 494 sites contain only gaps or ambiguous characters.\n",
      "Alignment has 18 sequences with 29903 columns, 193 distinct patterns\n",
      "109 parsimony-informative, 33 singleton sites, 29761 constant sites\n",
      "WARNING: Some sequence names are changed as follows:\n",
      "LR757995.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly__chromosome:_whole_genome -> LR757995.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly__chromosome__whole_genome\n",
      "LR757996.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly__chromosome:_whole_genome -> LR757996.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly__chromosome__whole_genome\n",
      "OL698718.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/MN-MDH-18236/2021_ORF1ab_polyprotein_(ORF1ab)__ORF1a_polyprotein_(ORF1ab)__surface_glycoprotein_(S)__ORF3a_protein_(ORF3a)__envelope_protein_(E)__membrane_glycoprotein_(M)__ORF6_protein_(ORF6)__ORF7a_protein_(ORF7a)__ORF7b_(ORF7b)__ORF8_protein_(ORF8)__nucleocapsid_phosphoprotein_(N)__and_ORF10_protein_(ORF10)_genes__complete_cds -> OL698718.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/MN-MDH-18236/2021_ORF1ab_polyprotein__ORF1ab___ORF1a_polyprotein__ORF1ab___surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___ORF6_protein__ORF6___ORF7a_protein__ORF7a___ORF7b__ORF7b___ORF8_protein__ORF8___nucleocapsid_phosphoprotein__N___and_ORF10_protein__ORF10__genes__complete_cds\n",
      "OL677199.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/CAN/ON-NML-249359/2021_ORF1ab_polyprotein_(ORF1ab)__ORF1a_polyprotein_(ORF1ab)__surface_glycoprotein_(S)__ORF3a_protein_(ORF3a)__envelope_protein_(E)__membrane_glycoprotein_(M)__ORF6_protein_(ORF6)__and_ORF7a_protein_(ORF7a)_genes__complete_cds;_ORF7b_gene__complete_sequence;_and_ORF8_protein_(ORF8)__nucleocapsid_phosphoprotein_(N)__and_ORF10_protein_(ORF10)_genes__complete_cds -> OL677199.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/CAN/ON-NML-249359/2021_ORF1ab_polyprotein__ORF1ab___ORF1a_polyprotein__ORF1ab___surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___ORF6_protein__ORF6___and_ORF7a_protein__ORF7a__genes__complete_cds__ORF7b_gene__complete_sequence__and_ORF8_protein__ORF8___nucleocapsid_phosphoprotein__N___and_ORF10_protein__ORF10__genes__complete_cds\n",
      "MZ914912.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/TG796484/2020_ORF1ab_polyprotein_(ORF1ab)__ORF1a_polyprotein_(ORF1ab)__surface_glycoprotein_(S)__ORF3a_protein_(ORF3a)__envelope_protein_(E)__membrane_glycoprotein_(M)__ORF6_protein_(ORF6)__ORF7a_protein_(ORF7a)__ORF7b_(ORF7b)__ORF8_protein_(ORF8)__nucleocapsid_phosphoprotein_(N)__and_ORF10_protein_(ORF10)_genes__complete_cds -> MZ914912.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/TG796484/2020_ORF1ab_polyprotein__ORF1ab___ORF1a_polyprotein__ORF1ab___surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___ORF6_protein__ORF6___ORF7a_protein__ORF7a___ORF7b__ORF7b___ORF8_protein__ORF8___nucleocapsid_phosphoprotein__N___and_ORF10_protein__ORF10__genes__complete_cds\n",
      "MZ916499.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/TG841289/2020_ORF1ab_polyprotein_(ORF1ab)_and_ORF1a_polyprotein_(ORF1ab)_genes__partial_cds;_and_surface_glycoprotein_(S)__ORF3a_protein_(ORF3a)__envelope_protein_(E)__membrane_glycoprotein_(M)__ORF6_protein_(ORF6)__ORF7a_protein_(ORF7a)__ORF7b_(ORF7b)__ORF8_protein_(ORF8)__nucleocapsid_phosphoprotein_(N)__and_ORF10_protein_(ORF10)_genes__complete_cds -> MZ916499.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/TG841289/2020_ORF1ab_polyprotein__ORF1ab__and_ORF1a_polyprotein__ORF1ab__genes__partial_cds__and_surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___ORF6_protein__ORF6___ORF7a_protein__ORF7a___ORF7b__ORF7b___ORF8_protein__ORF8___nucleocapsid_phosphoprotein__N___and_ORF10_protein__ORF10__genes__complete_cds\n",
      "MZ908464.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/TG769681/2020_ORF1ab_polyprotein_(ORF1ab)__ORF1a_polyprotein_(ORF1ab)__surface_glycoprotein_(S)__ORF3a_protein_(ORF3a)__envelope_protein_(E)__membrane_glycoprotein_(M)__and_ORF6_protein_(ORF6)_genes__complete_cds;_ORF7a_protein_(ORF7a)_and_ORF7b_(ORF7b)_genes__partial_cds;_and_ORF8_protein_(ORF8)__nucleocapsid_phosphoprotein_(N)__and_ORF10_protein_(ORF10)_genes__complete_cds -> MZ908464.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/TG769681/2020_ORF1ab_polyprotein__ORF1ab___ORF1a_polyprotein__ORF1ab___surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___and_ORF6_protein__ORF6__genes__complete_cds__ORF7a_protein__ORF7a__and_ORF7b__ORF7b__genes__partial_cds__and_ORF8_protein__ORF8___nucleocapsid_phosphoprotein__N___and_ORF10_protein__ORF10__genes__complete_cds\n",
      "MW580573.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/MD-MDH-0830/2021_ORF1ab_polyprotein_(ORF1ab)__ORF1a_polyprotein_(ORF1ab)__surface_glycoprotein_(S)__ORF3a_protein_(ORF3a)__envelope_protein_(E)__membrane_glycoprotein_(M)__ORF6_protein_(ORF6)__ORF7a_protein_(ORF7a)__ORF7b_(ORF7b)__ORF8_protein_(ORF8)__nucleocapsid_phosphoprotein_(N)__and_ORF10_protein_(ORF10)_genes__complete_cds -> MW580573.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/MD-MDH-0830/2021_ORF1ab_polyprotein__ORF1ab___ORF1a_polyprotein__ORF1ab___surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___ORF6_protein__ORF6___ORF7a_protein__ORF7a___ORF7b__ORF7b___ORF8_protein__ORF8___nucleocapsid_phosphoprotein__N___and_ORF10_protein__ORF10__genes__complete_cds\n",
      "MW991906.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/CA-CDC-FG-021330/2021_ORF1ab_polyprotein_(ORF1ab)_and_ORF1a_polyprotein_(ORF1ab)_genes__partial_cds;_surface_glycoprotein_(S)__ORF3a_protein_(ORF3a)__envelope_protein_(E)__membrane_glycoprotein_(M)__ORF6_protein_(ORF6)__ORF7a_protein_(ORF7a)__and_ORF7b_(ORF7b)_genes__complete_cds;_ORF8_gene__complete_sequence;_and_nucleocapsid_phosphoprotein_(N)_and_ORF10_protein_(ORF10)_genes__complete_cds -> MW991906.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/CA-CDC-FG-021330/2021_ORF1ab_polyprotein__ORF1ab__and_ORF1a_polyprotein__ORF1ab__genes__partial_cds__surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___ORF6_protein__ORF6___ORF7a_protein__ORF7a___and_ORF7b__ORF7b__genes__complete_cds__ORF8_gene__complete_sequence__and_nucleocapsid_phosphoprotein__N__and_ORF10_protein__ORF10__genes__complete_cds\n",
      "MW932027.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/MA-CDC-STM-000044850/2021_ORF1ab_polyprotein_(ORF1ab)__ORF1a_polyprotein_(ORF1ab)__surface_glycoprotein_(S)__ORF3a_protein_(ORF3a)__envelope_protein_(E)__membrane_glycoprotein_(M)__ORF6_protein_(ORF6)__ORF7a_protein_(ORF7a)__and_ORF7b_(ORF7b)_genes__complete_cds;_ORF8_gene__complete_sequence;_and_nucleocapsid_phosphoprotein_(N)_and_ORF10_protein_(ORF10)_genes__complete_cds -> MW932027.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/MA-CDC-STM-000044850/2021_ORF1ab_polyprotein__ORF1ab___ORF1a_polyprotein__ORF1ab___surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___ORF6_protein__ORF6___ORF7a_protein__ORF7a___and_ORF7b__ORF7b__genes__complete_cds__ORF8_gene__complete_sequence__and_nucleocapsid_phosphoprotein__N__and_ORF10_protein__ORF10__genes__complete_cds\n",
      "\n",
      "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   Gap/Ambiguity  Composition  p-value\n",
      "   1  NC_045512.2_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_Wuhan-Hu-1__complete_genome                                                                                                                                                                                                                                                                                                                                                                                                      1.65%    passed     99.98%\n",
      "   2  LR757995.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly__chromosome__whole_genome                                                                                                                                                                                                                                                                                                                                                                                                 1.65%    passed     99.98%\n",
      "   3  LR757996.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly__chromosome__whole_genome                                                                                                                                                                                                                                                                                                                                                                                                 1.65%    passed     99.98%\n",
      "   4  OL698718.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/MN-MDH-18236/2021_ORF1ab_polyprotein__ORF1ab___ORF1a_polyprotein__ORF1ab___surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___ORF6_protein__ORF6___ORF7a_protein__ORF7a___ORF7b__ORF7b___ORF8_protein__ORF8___nucleocapsid_phosphoprotein__N___and_ORF10_protein__ORF10__genes__complete_cds                                                                  3.28%    passed     99.65%\n",
      "   5  OL677199.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/CAN/ON-NML-249359/2021_ORF1ab_polyprotein__ORF1ab___ORF1a_polyprotein__ORF1ab___surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___ORF6_protein__ORF6___and_ORF7a_protein__ORF7a__genes__complete_cds__ORF7b_gene__complete_sequence__and_ORF8_protein__ORF8___nucleocapsid_phosphoprotein__N___and_ORF10_protein__ORF10__genes__complete_cds                     1.82%    passed     99.91%\n",
      "   6  OL672836.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/BEL/rega-20174/2021__complete_genome                                                                                                                                                                                                                                                                                                                                                                             1.78%    passed     99.96%\n",
      "   7  MZ914912.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/TG796484/2020_ORF1ab_polyprotein__ORF1ab___ORF1a_polyprotein__ORF1ab___surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___ORF6_protein__ORF6___ORF7a_protein__ORF7a___ORF7b__ORF7b___ORF8_protein__ORF8___nucleocapsid_phosphoprotein__N___and_ORF10_protein__ORF10__genes__complete_cds                                                                      2.69%    passed     99.99%\n",
      "   8  MZ916499.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/TG841289/2020_ORF1ab_polyprotein__ORF1ab__and_ORF1a_polyprotein__ORF1ab__genes__partial_cds__and_surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___ORF6_protein__ORF6___ORF7a_protein__ORF7a___ORF7b__ORF7b___ORF8_protein__ORF8___nucleocapsid_phosphoprotein__N___and_ORF10_protein__ORF10__genes__complete_cds                                            7.75%    passed     98.28%\n",
      "   9  MZ908464.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/TG769681/2020_ORF1ab_polyprotein__ORF1ab___ORF1a_polyprotein__ORF1ab___surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___and_ORF6_protein__ORF6__genes__complete_cds__ORF7a_protein__ORF7a__and_ORF7b__ORF7b__genes__partial_cds__and_ORF8_protein__ORF8___nucleocapsid_phosphoprotein__N___and_ORF10_protein__ORF10__genes__complete_cds                    5.93%    passed     96.00%\n",
      "  10  MW580573.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/MD-MDH-0830/2021_ORF1ab_polyprotein__ORF1ab___ORF1a_polyprotein__ORF1ab___surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___ORF6_protein__ORF6___ORF7a_protein__ORF7a___ORF7b__ORF7b___ORF8_protein__ORF8___nucleocapsid_phosphoprotein__N___and_ORF10_protein__ORF10__genes__complete_cds                                                                   2.26%    passed     99.99%\n",
      "  11  MW580574.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/MD-MDH-0831/2021__complete_genome                                                                                                                                                                                                                                                                                                                                                                            2.02%    passed     99.95%\n",
      "  12  MW580576.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/MD-MDH-0833/2021__complete_genome                                                                                                                                                                                                                                                                                                                                                                            1.98%    passed     99.93%\n",
      "  13  MW991906.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/CA-CDC-FG-021330/2021_ORF1ab_polyprotein__ORF1ab__and_ORF1a_polyprotein__ORF1ab__genes__partial_cds__surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___ORF6_protein__ORF6___ORF7a_protein__ORF7a___and_ORF7b__ORF7b__genes__complete_cds__ORF8_gene__complete_sequence__and_nucleocapsid_phosphoprotein__N__and_ORF10_protein__ORF10__genes__complete_cds    2.19%    passed     99.82%\n",
      "  14  MW931310.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/IN-CDC-STM-000045992/2021__complete_genome                                                                                                                                                                                                                                                                                                                                                                   1.68%    passed    100.00%\n",
      "  15  MW932027.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/MA-CDC-STM-000044850/2021_ORF1ab_polyprotein__ORF1ab___ORF1a_polyprotein__ORF1ab___surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___ORF6_protein__ORF6___ORF7a_protein__ORF7a___and_ORF7b__ORF7b__genes__complete_cds__ORF8_gene__complete_sequence__and_nucleocapsid_phosphoprotein__N__and_ORF10_protein__ORF10__genes__complete_cds                      1.70%    passed     99.98%\n",
      "  16  MW424864.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/CA-LACPHL-AF00051/2020__complete_genome                                                                                                                                                                                                                                                                                                                                                                      1.91%    passed     99.99%\n",
      "  17  MW453109.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/CA-LACPHL-AF00094/2020__complete_genome                                                                                                                                                                                                                                                                                                                                                                      2.13%    passed     99.99%\n",
      "  18  MW453110.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/CA-LACPHL-AF00093/2020__complete_genome                                                                                                                                                                                                                                                                                                                                                                      1.93%    passed     99.98%\n",
      "****  TOTAL                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                2.56%  0 sequences failed composition chi2 test (p-value<5%; df=3)\n",
      "NOTE: LR757996.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly__chromosome__whole_genome is identical to NC_045512.2_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_Wuhan-Hu-1__complete_genome but kept for subsequent analysis\n",
      "Creating fast initial parsimony tree by random order stepwise addition...\n",
      "0.003 seconds, parsimony score: 154 (based on 142 sites)\n",
      "\n",
      "NOTE: 0 MB RAM (0 GB) is required!\n",
      "WARNING: Number of threads seems too high for short alignments. Use -T AUTO to determine best number of threads.\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "1. Initial log-likelihood: -41361.570\n",
      "2. Current log-likelihood: -41335.547\n",
      "3. Current log-likelihood: -41330.199\n",
      "4. Current log-likelihood: -41321.132\n",
      "5. Current log-likelihood: -41320.973\n",
      "Optimal log-likelihood: -41320.963\n",
      "Rate parameters:  A-C: 1.00000  A-G: 4.58202  A-T: 1.00000  C-G: 1.00000  C-T: 4.58202  G-T: 1.00000\n",
      "Base frequencies:  A: 0.299  C: 0.183  G: 0.196  T: 0.322\n",
      "Parameters optimization took 5 rounds (0.015 sec)\n",
      "Computing ML distances based on estimated model parameters...\n",
      "Computing ML distances took 0.001753 sec (of wall-clock time) 0.005914 sec(of CPU time)\n",
      "Computing RapidNJ tree took 0.009550 sec (of wall-clock time) 0.009811 sec (of CPU time)\n",
      "Log-likelihood of RapidNJ tree: -41382.805\n",
      "--------------------------------------------------------------------\n",
      "|             INITIALIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Generating 98 parsimony trees... 0.079 second\n",
      "Computing log-likelihood of 98 initial trees ... 0.073 seconds\n",
      "Current best score: -41316.721\n",
      "\n",
      "Do NNI search on 20 best initial trees\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 1: -41316.682\n",
      "Iteration 10 / LogL: -41316.682 / Time: 0h:0m:0s\n",
      "Iteration 20 / LogL: -41316.682 / Time: 0h:0m:0s\n",
      "Finish initializing candidate tree set (1)\n",
      "Current best tree score: -41316.682 / CPU time: 0.246\n",
      "Number of iterations: 20\n",
      "--------------------------------------------------------------------\n",
      "|               OPTIMIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "UPDATE BEST LOG-LIKELIHOOD: -41316.682\n",
      "Iteration 30 / LogL: -41316.732 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 40 / LogL: -41316.719 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 50 / LogL: -41316.716 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 60 / LogL: -41341.534 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 70 / LogL: -41316.803 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "UPDATE BEST LOG-LIKELIHOOD: -41316.682\n",
      "Iteration 80 / LogL: -41327.750 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 90 / LogL: -41316.734 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "UPDATE BEST LOG-LIKELIHOOD: -41316.682\n",
      "Iteration 100 / LogL: -41316.803 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "TREE SEARCH COMPLETED AFTER 102 ITERATIONS / Time: 0h:0m:1s\n",
      "\n",
      "--------------------------------------------------------------------\n",
      "|                    FINALIZING TREE SEARCH                        |\n",
      "--------------------------------------------------------------------\n",
      "Performs final model parameters optimization\n",
      "Estimate model parameters (epsilon = 0.010)\n",
      "1. Initial log-likelihood: -41316.682\n",
      "Optimal log-likelihood: -41316.677\n",
      "Rate parameters:  A-C: 1.00000  A-G: 4.46795  A-T: 1.00000  C-G: 1.00000  C-T: 4.46795  G-T: 1.00000\n",
      "Base frequencies:  A: 0.299  C: 0.183  G: 0.196  T: 0.322\n",
      "Parameters optimization took 1 rounds (0.002 sec)\n",
      "BEST SCORE FOUND : -41316.677\n",
      "Total tree length: 0.005\n",
      "\n",
      "Total number of iterations: 102\n",
      "CPU time used for tree search: 4.260 sec (0h:0m:4s)\n",
      "Wall-clock time used for tree search: 1.142 sec (0h:0m:1s)\n",
      "Total CPU time used: 4.401 sec (0h:0m:4s)\n",
      "Total wall-clock time used: 1.190 sec (0h:0m:1s)\n",
      "\n",
      "Analysis results written to: \n",
      "  IQ-TREE report:                sarscov2_tree.iqtree\n",
      "  Maximum-likelihood tree:       sarscov2_tree.treefile\n",
      "  Likelihood distances:          sarscov2_tree.mldist\n",
      "  Screen log file:               sarscov2_tree.log\n",
      "\n",
      "Date and Time: Fri Dec  3 23:53:06 2021\n"
     ]
    }
   ],
   "source": [
    "#run iqtree with threads = $CPU variable, if you exclude the -m it will do a phylogenetic model search before tree search\n",
    "!iqtree -s sequences.aln.fasta -nt $CPU -m HKY --prefix sarscov2_tree --redo-tree"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7197dd4",
   "metadata": {},
   "source": [
    "### Visualize the tree with toytree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "cef2ba18",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Define the tree file\n",
    "tre = toytree.tree('sarscov2_tree.treefile')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "842af165",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"toyplot\" id=\"t5b24751d28514045a6c48e279637cb85\" style=\"text-align:center\"><svg class=\"toyplot-canvas-Canvas\" height=\"324.0px\" id=\"t854aace90f344035b0524662eb5e8bd5\" preserveAspectRatio=\"xMidYMid meet\" style=\"background-color:transparent;border-color:#292724;border-style:none;border-width:1.0;fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:Helvetica;font-size:12px;opacity:1.0;stroke:rgb(16.1%,15.3%,14.1%);stroke-opacity:1.0;stroke-width:1.0\" viewBox=\"0 0 500.0 324.0\" width=\"500.0px\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:toyplot=\"http://www.sandia.gov/toyplot\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><g class=\"toyplot-coordinates-Cartesian\" id=\"t947bd6e6296d45e6b2a424dfe8761113\"><clipPath id=\"t98314000403f403e922aaabec48413a5\"><rect height=\"264.0\" width=\"440.0\" x=\"30.0\" y=\"30.0\"></rect></clipPath><g clip-path=\"url(#t98314000403f403e922aaabec48413a5)\"><g class=\"toytree-mark-Toytree\" id=\"t4c087087ca4f4a03822c7a82b839296d\"><g class=\"toytree-Edges\" style=\"fill:none;stroke:rgb(14.9%,14.9%,14.9%);stroke-linecap:round;stroke-opacity:1;stroke-width:2\"><path d=\"M 50.1 104.6 L 50.1 70.2 L 62.5 70.2\" id=\"34,33\"></path><path d=\"M 50.1 104.6 L 50.1 139.1 L 62.5 139.1\" id=\"34,32\"></path><path d=\"M 62.5 70.2 L 62.5 79.1 L 62.5 79.1\" id=\"33,31\"></path><path d=\"M 62.5 139.1 L 62.5 105.7 L 69.0 105.7\" id=\"32,30\"></path><path d=\"M 62.5 139.1 L 62.5 172.4 L 63.2 172.4\" id=\"32,29\"></path><path d=\"M 69.0 105.7 L 69.0 114.6 L 71.0 114.6\" id=\"30,28\"></path><path d=\"M 63.2 172.4 L 63.2 141.3 L 64.7 141.3\" id=\"29,27\"></path><path d=\"M 63.2 172.4 L 63.2 203.5 L 64.2 203.5\" id=\"29,26\"></path><path d=\"M 64.7 141.3 L 64.7 150.2 L 64.7 150.2\" id=\"27,25\"></path><path d=\"M 64.2 203.5 L 64.2 176.8 L 68.2 176.8\" id=\"26,24\"></path><path d=\"M 64.2 203.5 L 64.2 230.1 L 64.7 230.1\" id=\"26,23\"></path><path d=\"M 68.2 176.8 L 68.2 185.7 L 68.8 185.7\" id=\"24,22\"></path><path d=\"M 64.7 230.1 L 64.7 212.3 L 72.3 212.3\" id=\"23,21\"></path><path d=\"M 64.7 230.1 L 64.7 247.9 L 69.7 247.9\" id=\"23,20\"></path><path d=\"M 72.3 212.3 L 72.3 221.2 L 72.3 221.2\" id=\"21,19\"></path><path d=\"M 69.7 247.9 L 69.7 256.8 L 73.3 256.8\" id=\"20,18\"></path><path d=\"M 62.5 70.2 L 62.5 61.3 L 63.5 61.3\" id=\"33,17\"></path><path d=\"M 62.5 79.1 L 62.5 73.2 L 62.5 73.2\" id=\"31,16\"></path><path d=\"M 62.5 79.1 L 62.5 85.0 L 62.5 85.0\" id=\"31,15\"></path><path d=\"M 69.0 105.7 L 69.0 96.9 L 73.4 96.9\" id=\"30,14\"></path><path d=\"M 71.0 114.6 L 71.0 108.7 L 77.7 108.7\" id=\"28,13\"></path><path d=\"M 71.0 114.6 L 71.0 120.5 L 74.8 120.5\" id=\"28,12\"></path><path d=\"M 64.7 141.3 L 64.7 132.4 L 65.7 132.4\" id=\"27,11\"></path><path d=\"M 64.7 150.2 L 64.7 144.2 L 64.7 144.2\" id=\"25,10\"></path><path d=\"M 64.7 150.2 L 64.7 156.1 L 64.7 156.1\" id=\"25,9\"></path><path d=\"M 68.2 176.8 L 68.2 167.9 L 68.3 167.9\" id=\"24,8\"></path><path d=\"M 68.8 185.7 L 68.8 179.8 L 68.8 179.8\" id=\"22,7\"></path><path d=\"M 68.8 185.7 L 68.8 191.6 L 68.8 191.6\" id=\"22,6\"></path><path d=\"M 72.3 212.3 L 72.3 203.5 L 72.8 203.5\" id=\"21,5\"></path><path d=\"M 72.3 221.2 L 72.3 215.3 L 72.3 215.3\" id=\"19,4\"></path><path d=\"M 72.3 221.2 L 72.3 227.1 L 72.3 227.1\" id=\"19,3\"></path><path d=\"M 69.7 247.9 L 69.7 239.0 L 71.8 239.0\" id=\"20,2\"></path><path d=\"M 73.3 256.8 L 73.3 250.8 L 73.3 250.8\" id=\"18,1\"></path><path d=\"M 73.3 256.8 L 73.3 262.7 L 73.3 262.7\" id=\"18,0\"></path></g><g class=\"toytree-AlignEdges\" style=\"stroke:rgb(66.3%,66.3%,66.3%);stroke-dasharray:2, 4;stroke-linecap:round;stroke-opacity:1.0;stroke-width:2\"><path d=\"M 77.7 262.7 L 73.3 262.7\"></path><path d=\"M 77.7 250.8 L 73.3 250.8\"></path><path d=\"M 77.7 239.0 L 71.8 239.0\"></path><path d=\"M 77.7 227.1 L 72.3 227.1\"></path><path d=\"M 77.7 215.3 L 72.3 215.3\"></path><path d=\"M 77.7 203.5 L 72.8 203.5\"></path><path d=\"M 77.7 191.6 L 68.8 191.6\"></path><path d=\"M 77.7 179.8 L 68.8 179.8\"></path><path d=\"M 77.7 167.9 L 68.3 167.9\"></path><path d=\"M 77.7 156.1 L 64.7 156.1\"></path><path d=\"M 77.7 144.2 L 64.7 144.2\"></path><path d=\"M 77.7 132.4 L 65.7 132.4\"></path><path d=\"M 77.7 120.5 L 74.8 120.5\"></path><path d=\"M 77.7 108.7 L 77.7 108.7\"></path><path d=\"M 77.7 96.9 L 73.4 96.9\"></path><path d=\"M 77.7 85.0 L 62.5 85.0\"></path><path d=\"M 77.7 73.2 L 62.5 73.2\"></path><path d=\"M 77.7 61.3 L 63.5 61.3\"></path></g><g class=\"toytree-TipLabels\" style=\"fill:rgb(14.9%,14.9%,14.9%);fill-opacity:1.0;font-family:helvetica;font-size:11px;font-weight:normal;stroke:none;white-space:pre\"><g transform=\"translate(77.73,262.67)rotate(0)\"><text style=\"\" x=\"15.00\" y=\"2.81\">MW453110.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/CA-LACPHL-AF00093/2020__complete_genome</text></g><g transform=\"translate(77.73,250.83)rotate(0)\"><text style=\"\" x=\"15.00\" y=\"2.81\">MW453109.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/CA-LACPHL-AF00094/2020__complete_genome</text></g><g transform=\"translate(77.73,238.99)rotate(0)\"><text style=\"\" x=\"15.00\" y=\"2.81\">MW424864.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/CA-LACPHL-AF00051/2020__complete_genome</text></g><g transform=\"translate(77.73,227.14)rotate(0)\"><text style=\"\" x=\"15.00\" y=\"2.81\">MW580576.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/MD-MDH-0833/2021__complete_genome</text></g><g transform=\"translate(77.73,215.30)rotate(0)\"><text style=\"\" x=\"15.00\" y=\"2.81\">MW580574.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/MD-MDH-0831/2021__complete_genome</text></g><g transform=\"translate(77.73,203.45)rotate(0)\"><text style=\"\" x=\"15.00\" y=\"2.81\">MW580573.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/MD-MDH-0830/2021_ORF1ab_polyprotein__ORF1ab___ORF1a_polyprotein__ORF1ab___surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___ORF6_protein__ORF6___ORF7a_protein__ORF7a___ORF7b__ORF7b___ORF8_protein__ORF8___nucleocapsid_phosphoprotein__N___and_ORF10_protein__ORF10__genes__complete_cds</text></g><g transform=\"translate(77.73,191.61)rotate(0)\"><text style=\"\" x=\"15.00\" y=\"2.81\">MZ908464.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/TG769681/2020_ORF1ab_polyprotein__ORF1ab___ORF1a_polyprotein__ORF1ab___surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___and_ORF6_protein__ORF6__genes__complete_cds__ORF7a_protein__ORF7a__and_ORF7b__ORF7b__genes__partial_cds__and_ORF8_protein__ORF8___nucleocapsid_phosphoprotein__N___and_ORF10_protein__ORF10__genes__complete_cds</text></g><g transform=\"translate(77.73,179.77)rotate(0)\"><text style=\"\" x=\"15.00\" y=\"2.81\">MZ914912.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/TG796484/2020_ORF1ab_polyprotein__ORF1ab___ORF1a_polyprotein__ORF1ab___surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___ORF6_protein__ORF6___ORF7a_protein__ORF7a___ORF7b__ORF7b___ORF8_protein__ORF8___nucleocapsid_phosphoprotein__N___and_ORF10_protein__ORF10__genes__complete_cds</text></g><g transform=\"translate(77.73,167.92)rotate(0)\"><text style=\"\" x=\"15.00\" y=\"2.81\">MZ916499.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/TG841289/2020_ORF1ab_polyprotein__ORF1ab__and_ORF1a_polyprotein__ORF1ab__genes__partial_cds__and_surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___ORF6_protein__ORF6___ORF7a_protein__ORF7a___ORF7b__ORF7b___ORF8_protein__ORF8___nucleocapsid_phosphoprotein__N___and_ORF10_protein__ORF10__genes__complete_cds</text></g><g transform=\"translate(77.73,156.08)rotate(0)\"><text style=\"\" x=\"15.00\" y=\"2.81\">LR757996.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly__chromosome__whole_genome</text></g><g transform=\"translate(77.73,144.23)rotate(0)\"><text style=\"\" x=\"15.00\" y=\"2.81\">NC_045512.2_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_Wuhan-Hu-1__complete_genome</text></g><g transform=\"translate(77.73,132.39)rotate(0)\"><text style=\"\" x=\"15.00\" y=\"2.81\">LR757995.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly__chromosome__whole_genome</text></g><g transform=\"translate(77.73,120.55)rotate(0)\"><text style=\"\" x=\"15.00\" y=\"2.81\">MW932027.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/MA-CDC-STM-000044850/2021_ORF1ab_polyprotein__ORF1ab___ORF1a_polyprotein__ORF1ab___surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___ORF6_protein__ORF6___ORF7a_protein__ORF7a___and_ORF7b__ORF7b__genes__complete_cds__ORF8_gene__complete_sequence__and_nucleocapsid_phosphoprotein__N__and_ORF10_protein__ORF10__genes__complete_cds</text></g><g transform=\"translate(77.73,108.70)rotate(0)\"><text style=\"\" x=\"15.00\" y=\"2.81\">MW991906.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/CA-CDC-FG-021330/2021_ORF1ab_polyprotein__ORF1ab__and_ORF1a_polyprotein__ORF1ab__genes__partial_cds__surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___ORF6_protein__ORF6___ORF7a_protein__ORF7a___and_ORF7b__ORF7b__genes__complete_cds__ORF8_gene__complete_sequence__and_nucleocapsid_phosphoprotein__N__and_ORF10_protein__ORF10__genes__complete_cds</text></g><g transform=\"translate(77.73,96.86)rotate(0)\"><text style=\"\" x=\"15.00\" y=\"2.81\">MW931310.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/IN-CDC-STM-000045992/2021__complete_genome</text></g><g transform=\"translate(77.73,85.01)rotate(0)\"><text style=\"\" x=\"15.00\" y=\"2.81\">OL672836.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/BEL/rega-20174/2021__complete_genome</text></g><g transform=\"translate(77.73,73.17)rotate(0)\"><text style=\"\" x=\"15.00\" y=\"2.81\">OL677199.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/CAN/ON-NML-249359/2021_ORF1ab_polyprotein__ORF1ab___ORF1a_polyprotein__ORF1ab___surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___ORF6_protein__ORF6___and_ORF7a_protein__ORF7a__genes__complete_cds__ORF7b_gene__complete_sequence__and_ORF8_protein__ORF8___nucleocapsid_phosphoprotein__N___and_ORF10_protein__ORF10__genes__complete_cds</text></g><g transform=\"translate(77.73,61.33)rotate(0)\"><text style=\"\" x=\"15.00\" y=\"2.81\">OL698718.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2/human/USA/MN-MDH-18236/2021_ORF1ab_polyprotein__ORF1ab___ORF1a_polyprotein__ORF1ab___surface_glycoprotein__S___ORF3a_protein__ORF3a___envelope_protein__E___membrane_glycoprotein__M___ORF6_protein__ORF6___ORF7a_protein__ORF7a___ORF7b__ORF7b___ORF8_protein__ORF8___nucleocapsid_phosphoprotein__N___and_ORF10_protein__ORF10__genes__complete_cds</text></g></g></g></g></g></svg><div class=\"toyplot-behavior\"><script>(function()\n",
       "{\n",
       "var modules={};\n",
       "})();</script></div></div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#draw the tree\n",
    "rtre = tre.root(wildcard=\"OL\")\n",
    "rtre.draw(tip_labels_align=True);"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52d9389f",
   "metadata": {},
   "source": [
    "You can also visualize the tree by downloading it and opening in figtree."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88457512",
   "metadata": {},
   "source": [
    "And that is all! You now know how to run workflows in notebooks in Cloud Lab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e417cb1a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "environment": {
   "kernel": "python3",
   "name": "r-cpu.4-1.m87",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/r-cpu.4-1:m87"
  },
  "kernelspec": {
   "display_name": "conda_amazonei_mxnet_p36",
   "language": "python",
   "name": "conda_amazonei_mxnet_p36"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
